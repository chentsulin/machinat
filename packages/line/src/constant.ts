export const CHANNEL_REQUEST_GETTER = Symbol(
  'channel_request_getter.line.machinat'
);
export const BULK_REQUEST_GETTER = Symbol('bulk_request_getter.line.machinat');

/** @internal */
export const LINE = 'line' as const;

/** @internal */
export const PATH_REPLY = 'v2/bot/message/reply';

/** @internal */
export const PATH_PUSH = 'v2/bot/message/push';

/** @internal */
export const PATH_MULTICAST = 'v2/bot/message/multicast';

/** @internal */
export const PATH_RICHMENU = 'v2/bot/richmenu';

export enum LiffContextOs {
  Ios,
  Android,
  Web,
}
