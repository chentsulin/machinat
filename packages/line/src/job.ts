/** @internal */ /** */
import type { DispatchableSegment } from '@machinat/core/engine/types';
import LineChannel from './channel';
import type {
  LineSegmentValue,
  LineMessageSegmentValue,
  LineJob,
} from './types';
import {
  PATH_PUSH,
  PATH_REPLY,
  PATH_MULTICAST,
  CHANNEL_REQUEST_GETTER,
  BULK_REQUEST_GETTER,
} from './constant';
import { isMessageValue } from './utils';

const createMessageJob = (
  channel: LineChannel,
  messages: LineMessageSegmentValue[],
  replyToken: void | string
): LineJob => ({
  method: 'POST',
  path: replyToken ? PATH_REPLY : PATH_PUSH,
  executionKey: channel.uid,
  body: replyToken
    ? { replyToken: replyToken as string, messages }
    : { to: channel.id, messages },
});

export const chatJobsMaker = (replyToken: void | string) => {
  let messagingJobsCount = 0;

  return (
    channel: LineChannel,
    segments: DispatchableSegment<LineSegmentValue>[]
  ): LineJob[] => {
    const jobs: LineJob[] = [];
    let messagesBuffer: LineMessageSegmentValue[] = [];

    for (let i = 0; i < segments.length; i += 1) {
      const { value } = segments[i];

      if (isMessageValue(value)) {
        messagesBuffer.push(
          typeof value === 'string'
            ? { type: 'text', text: value as string }
            : value
        );

        // flush messages buffer if accumlated to 5 or at the end of loop
        if (messagesBuffer.length === 5 || i === segments.length - 1) {
          jobs.push(createMessageJob(channel, messagesBuffer, replyToken));
          messagesBuffer = [];
          messagingJobsCount += 1;
        }
      } else {
        // push buffered messages first
        if (messagesBuffer.length > 0) {
          jobs.push(createMessageJob(channel, messagesBuffer, replyToken));
          messagesBuffer = [];
          messagingJobsCount += 1;
        }

        // get dynamic api request
        const { method, path, body } = value[CHANNEL_REQUEST_GETTER](channel);
        jobs.push({
          method,
          path,
          executionKey: channel.uid,
          body,
        });
      }
    }

    if (replyToken && messagingJobsCount > 1) {
      throw new RangeError(
        'more then 1 messaging request rendered while using replyToken'
      );
    }

    return jobs;
  };
};

const MULITCAST_EXECUTION_KEY = '$$_multicast_$$';

export const multicastJobsMaker = (targets: string[]) => (
  _: null,
  segments: DispatchableSegment<LineSegmentValue>[]
) => {
  const jobs: LineJob[] = [];
  let messages: LineSegmentValue[] = [];

  for (let i = 0; i < segments.length; i += 1) {
    const { value } = segments[i];

    if (isMessageValue(value)) {
      messages.push(
        typeof value === 'string'
          ? { type: 'text', text: value as string }
          : value
      );

      // flush messages buffer if accumlated to 5 or at the end of loop
      if (messages.length === 5 || i === segments.length - 1) {
        jobs.push({
          method: 'POST',
          path: PATH_MULTICAST,
          body: { to: targets, messages },
          executionKey: MULITCAST_EXECUTION_KEY,
        });
        messages = [];
      }
    } else {
      // push buffered messages first
      if (messages.length > 0) {
        jobs.push({
          method: 'POST',
          path: PATH_MULTICAST,
          body: { to: targets, messages },
          executionKey: MULITCAST_EXECUTION_KEY,
        });
        messages = [];
      }

      // get dynamic api request
      const { method, path, body } = value[BULK_REQUEST_GETTER](targets);
      jobs.push({
        method,
        path,
        executionKey: MULITCAST_EXECUTION_KEY,
        body,
      });
    }
  }

  return jobs;
};
