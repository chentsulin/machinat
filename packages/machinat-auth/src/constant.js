// @flow
import { namedInterface } from '@machinat/core/service';
import type { ServerAuthorizer, AuthModuleConfigs } from './types';

export const STATE_COOKIE_KEY = 'machinat_auth_state';
export const TOKEN_COOKIE_KEY = 'machinat_auth_token';
export const SIGNATURE_COOKIE_KEY = 'machinat_auth_signature';
export const ERROR_COOKIE_KEY = 'machinat_auth_error';

export const AUTH_SERVER_AUTHORIZERS_I = namedInterface<
  ServerAuthorizer<any, any>[]
>({
  name: 'ServerAuthorizersList',
  multi: true,
});

export const AUTH_MODULE_CONFIGS_I = namedInterface<AuthModuleConfigs>({
  name: 'AuthModuleConfigs',
});
