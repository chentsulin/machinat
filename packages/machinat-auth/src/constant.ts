import { makeInterface } from '@machinat/core/service';
import type { ServerAuthorizer, AuthModuleConfigs } from './types';

export const STATE_COOKIE_KEY = 'machinat_auth_state';
export const TOKEN_COOKIE_KEY = 'machinat_auth_token';
export const SIGNATURE_COOKIE_KEY = 'machinat_auth_signature';
export const ERROR_COOKIE_KEY = 'machinat_auth_error';

export const SERVER_AUTHORIZERS_I = makeInterface<ServerAuthorizer<any, any>[]>(
  { name: 'ServerAuthorizersList', multi: true }
);

export const MODULE_CONFIGS_I = makeInterface<AuthModuleConfigs>({
  name: 'AuthModuleConfigs',
});
