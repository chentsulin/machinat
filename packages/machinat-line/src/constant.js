// @flow
export const LINE = 'line';
export const LINE_NAMESPACE = 'Line';

export const LINE_NATIVE_TYPE = Symbol.for('machinat.line.native');

export const ENTRY_REPLY = 'v2/bot/message/reply';
export const ENTRY_PUSH = 'v2/bot/message/push';
export const ENTRY_MULTICAST = 'v2/bot/message/multicast';
