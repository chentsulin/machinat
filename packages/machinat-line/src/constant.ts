export const LINE = 'line';

export const CHANNEL_API_CALL_GETTER = Symbol(
  'line.segment.channel_api_call_getter'
);
export const BULK_API_CALL_GETTER = Symbol('line.segment.bulk_api_call_getter');

export const PATH_REPLY = 'v2/bot/message/reply';
export const PATH_PUSH = 'v2/bot/message/push';
export const PATH_MULTICAST = 'v2/bot/message/multicast';
export const PATH_RICHMENU = 'v2/bot/richmenu';
