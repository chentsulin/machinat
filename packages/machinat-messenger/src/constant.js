// @flow
export const MESSENGER = 'messenger';
export const MESSENGER_NAMESPACE = 'Messenger';
export const MESSENGER_NATIVE_TYPE = Symbol.for('machinat.messenger.native');

export const ATTACHED_FILE_DATA = Symbol(
  'machinat.messenger.job.attached_file_data'
);

export const ATTACHED_FILE_INFO = Symbol(
  'machinat.messenger.job.attached_file_info'
);

export const ENTRY_MESSAGES = 'me/messages';

export const ENTRY_PASS_THREAD_CONTROL = 'me/pass_thread_control';
export const ENTRY_REQUEST_THREAD_CONTROL = 'me/request_thread_control';
export const ENTRY_TAKE_THREAD_CONTROL = 'me/take_thread_control';

export const ENTRY_MESSAGE_CREATIVES = 'me/message_creatives';
export const ENTRY_BROADCAST_MESSAGES = 'me/broadcast_messages';
export const ENTRY_CUSTOM_LABELS = 'me/custom_labels';
