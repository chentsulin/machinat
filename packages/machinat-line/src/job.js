// @flow
import invariant from 'invariant';
import { formatNode } from 'machinat-utility';

import type { SegmentWithoutPause } from 'machinat-base/types';
import type LineChannel from './channel';
import type {
  LineSegmentValue,
  LineComponent,
  LineJob,
  LineSendOptions,
} from './types';
import { ENTRY_PUSH, ENTRY_REPLY, ENTRY_MULTICAST } from './constant';

const isMessagesEntry = node =>
  typeof node !== 'object' ||
  typeof node.type !== 'function' ||
  node.type.$$getEntry === undefined;

export const createChatJobs = (
  channel: LineChannel,
  segments: SegmentWithoutPause<LineSegmentValue, LineComponent>[],
  options?: LineSendOptions
) => {
  const replyToken = options && options.replyToken;

  const jobs: LineJob[] = [];
  let messagesBuffer: LineSegmentValue[] = [];

  for (let i = 0; i < segments.length; i += 1) {
    const { node, value } = segments[i];

    const isMesssage = isMessagesEntry(node);

    // push message to buffer
    if (isMesssage) {
      messagesBuffer.push(
        typeof value === 'string'
          ? { type: 'text', text: (value: string) }
          : value
      );
    }

    // if non-message met, loop ended or messagesBuffer accumulated up to 5
    // move buffered messages into job
    if (
      messagesBuffer.length > 0 &&
      (!isMesssage || messagesBuffer.length === 5 || i === segments.length - 1)
    ) {
      jobs.push({
        method: 'POST',
        entry: replyToken ? ENTRY_REPLY : ENTRY_PUSH,
        channelUid: channel.uid,
        body: replyToken
          ? { replyToken: (replyToken: string), messages: messagesBuffer }
          : { to: (channel.sourceId: string), messages: messagesBuffer },
      });
      messagesBuffer = [];
    }

    // if non message met, use value as body directly and get dynamic entry
    if (
      !isMesssage /* :: && // for refinement, this equals to:
      typeof node === 'object' &&
      typeof node.type === 'function' &&
      node.type.$$getEntry !== undefined */
    ) {
      jobs.push({
        method: 'POST',
        entry: node.type.$$getEntry(channel, value),
        channelUid: channel.uid,
        body: value,
      });
    }
  }

  return jobs;
};

export const createMulticastJobs = (
  targets: string[],
  segments: SegmentWithoutPause<LineSegmentValue, LineComponent>[]
) => {
  const jobs: LineJob[] = [];
  let messages: LineSegmentValue[] = [];

  for (let i = 0; i < segments.length; i += 1) {
    const { node, value } = segments[i];

    invariant(
      isMessagesEntry(node),
      `${formatNode(node)} is invalid to be delivered in multicast`
    );

    messages.push(
      typeof value === 'string'
        ? { type: 'text', text: (value: string) }
        : value
    );

    if (messages.length === 5 || i === segments.length - 1) {
      jobs.push({
        method: 'POST',
        entry: ENTRY_MULTICAST,
        body: { to: targets, messages },
      });
      messages = [];
    }
  }

  return jobs;
};
