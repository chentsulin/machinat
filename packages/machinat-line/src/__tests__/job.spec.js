import moxy from 'moxy';
import Machinat from '@machinat/core';
import { CHANNEL_API_CALL_GETTER, BULK_API_CALL_GETTER } from '../constant';
import LineChannel from '../channel';
import { chatJobsMaker, multicastJobsMaker } from '../job';

const Foo = () => {};
const Bar = () => {};
const Baz = () => {};

const dynamicAPICaller = moxy({
  [CHANNEL_API_CALL_GETTER]: () => ({
    method: 'POST',
    path: 'some/channel/api',
    body: { do: 'something' },
  }),
  [BULK_API_CALL_GETTER]: () => ({
    method: 'POST',
    path: 'some/bulk/api',
    body: { bulk: 'do something' },
  }),
});

const segments = [
  { node: <Foo />, value: { id: 0 } },
  { node: <Foo />, value: { id: 1 } },
  { node: <Foo />, value: { id: 2 } },
  { node: <Foo />, value: { id: 3 } },
  { node: <Foo />, value: { id: 4 } },
  { node: <Foo />, value: { id: 5 } },
  { node: <Foo />, value: { id: 6 } },
  { node: <Bar />, value: { id: 7, ...dynamicAPICaller } },
  { node: <Foo />, value: { id: 8 } },
  { node: <Baz />, value: { id: 9, ...dynamicAPICaller } },
];

beforeEach(() => {
  dynamicAPICaller.mock.clear();
});

describe('chatJobsMaker()', () => {
  it('make push jobs', () => {
    const channel = new LineChannel('_LINE_CHANNEL_ID_', {
      type: 'user',
      userId: 'john',
    });

    const jobs = chatJobsMaker()(channel, segments);

    expect(jobs).toMatchInlineSnapshot(`
      Array [
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "id": 0,
              },
              Object {
                "id": 1,
              },
              Object {
                "id": 2,
              },
              Object {
                "id": 3,
              },
              Object {
                "id": 4,
              },
            ],
            "to": "john",
          },
          "channelUid": "line._LINE_CHANNEL_ID_.user.john",
          "method": "POST",
          "path": "v2/bot/message/push",
        },
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "id": 5,
              },
              Object {
                "id": 6,
              },
            ],
            "to": "john",
          },
          "channelUid": "line._LINE_CHANNEL_ID_.user.john",
          "method": "POST",
          "path": "v2/bot/message/push",
        },
        Object {
          "body": Object {
            "do": "something",
          },
          "channelUid": "line._LINE_CHANNEL_ID_.user.john",
          "method": "POST",
          "path": "some/channel/api",
        },
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "id": 8,
              },
            ],
            "to": "john",
          },
          "channelUid": "line._LINE_CHANNEL_ID_.user.john",
          "method": "POST",
          "path": "v2/bot/message/push",
        },
        Object {
          "body": Object {
            "do": "something",
          },
          "channelUid": "line._LINE_CHANNEL_ID_.user.john",
          "method": "POST",
          "path": "some/channel/api",
        },
      ]
    `);

    expect(
      dynamicAPICaller[CHANNEL_API_CALL_GETTER].mock
    ).toHaveBeenCalledTimes(2);
    expect(dynamicAPICaller[CHANNEL_API_CALL_GETTER].mock).toHaveBeenCalledWith(
      channel
    );
  });

  test('make reply job', () => {
    const channel = new LineChannel('_LINE_CHANNEL_ID_', {
      type: 'user',
      userId: 'john',
    });

    const jobs = chatJobsMaker('__REPLY_TOKEN__')(
      channel,
      segments.slice(0, 5)
    );

    expect(jobs).toMatchInlineSnapshot(`
      Array [
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "id": 0,
              },
              Object {
                "id": 1,
              },
              Object {
                "id": 2,
              },
              Object {
                "id": 3,
              },
              Object {
                "id": 4,
              },
            ],
            "replyToken": "__REPLY_TOKEN__",
          },
          "channelUid": "line._LINE_CHANNEL_ID_.user.john",
          "method": "POST",
          "path": "v2/bot/message/reply",
        },
      ]
    `);
  });

  it('throw if multiple messaging job made when reply', () => {
    const channel = new LineChannel('_LINE_CHANNEL_ID_', {
      type: 'user',
      userId: 'john',
    });

    expect(() =>
      chatJobsMaker('__REPLY_TOKEN__')(channel, segments)
    ).toThrowErrorMatchingInlineSnapshot(
      `"more then 1 messaging request rendered while using replyToken"`
    );
  });
});

describe('multicastJobsMaker()', () => {
  it('create job of multicast api', () => {
    const jobs = multicastJobsMaker(['foo', 'bar', 'baz'])(null, segments);

    expect(jobs).toMatchInlineSnapshot(`
      Array [
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "id": 0,
              },
              Object {
                "id": 1,
              },
              Object {
                "id": 2,
              },
              Object {
                "id": 3,
              },
              Object {
                "id": 4,
              },
            ],
            "to": Array [
              "foo",
              "bar",
              "baz",
            ],
          },
          "method": "POST",
          "path": "v2/bot/message/multicast",
        },
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "id": 5,
              },
              Object {
                "id": 6,
              },
            ],
            "to": Array [
              "foo",
              "bar",
              "baz",
            ],
          },
          "method": "POST",
          "path": "v2/bot/message/multicast",
        },
        Object {
          "body": Object {
            "bulk": "do something",
          },
          "channelUid": undefined,
          "method": "POST",
          "path": "some/bulk/api",
        },
        Object {
          "body": Object {
            "messages": Array [
              Object {
                "id": 8,
              },
            ],
            "to": Array [
              "foo",
              "bar",
              "baz",
            ],
          },
          "method": "POST",
          "path": "v2/bot/message/multicast",
        },
        Object {
          "body": Object {
            "bulk": "do something",
          },
          "channelUid": undefined,
          "method": "POST",
          "path": "some/bulk/api",
        },
      ]
    `);

    expect(jobs[0].channelUid).toBe(undefined);
    expect(jobs[0].channelUid).toBe(undefined);

    expect(dynamicAPICaller[BULK_API_CALL_GETTER].mock).toHaveBeenCalledTimes(
      2
    );
    expect(dynamicAPICaller[BULK_API_CALL_GETTER].mock).toHaveBeenCalledWith([
      'foo',
      'bar',
      'baz',
    ]);
  });
});
