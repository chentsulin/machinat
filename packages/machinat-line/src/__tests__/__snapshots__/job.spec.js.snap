// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createChatJobs() when useReplyAPI 1`] = `
Array [
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 0,
        },
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
        Object {
          "id": 4,
        },
      ],
      "replyToken": "__REPLY_TOKEN__",
    },
    "channelUid": "line:*:user:john",
    "entry": "v2/bot/message/reply",
    "method": "POST",
  },
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 5,
        },
        Object {
          "id": 6,
        },
      ],
      "replyToken": "__REPLY_TOKEN__",
    },
    "channelUid": "line:*:user:john",
    "entry": "v2/bot/message/reply",
    "method": "POST",
  },
  Object {
    "body": Object {
      "id": 7,
    },
    "channelUid": "line:*:user:john",
    "entry": "bar",
    "method": "POST",
  },
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 8,
        },
      ],
      "replyToken": "__REPLY_TOKEN__",
    },
    "channelUid": "line:*:user:john",
    "entry": "v2/bot/message/reply",
    "method": "POST",
  },
  Object {
    "body": Object {
      "id": 9,
    },
    "channelUid": "line:*:user:john",
    "entry": "baz",
    "method": "POST",
  },
]
`;

exports[`createChatJobs() work 1`] = `
Array [
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 0,
        },
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
        Object {
          "id": 4,
        },
      ],
      "to": "john",
    },
    "channelUid": "line:*:user:john",
    "entry": "v2/bot/message/push",
    "method": "POST",
  },
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 5,
        },
        Object {
          "id": 6,
        },
      ],
      "to": "john",
    },
    "channelUid": "line:*:user:john",
    "entry": "v2/bot/message/push",
    "method": "POST",
  },
  Object {
    "body": Object {
      "id": 7,
    },
    "channelUid": "line:*:user:john",
    "entry": "bar",
    "method": "POST",
  },
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 8,
        },
      ],
      "to": "john",
    },
    "channelUid": "line:*:user:john",
    "entry": "v2/bot/message/push",
    "method": "POST",
  },
  Object {
    "body": Object {
      "id": 9,
    },
    "channelUid": "line:*:user:john",
    "entry": "baz",
    "method": "POST",
  },
]
`;

exports[`createMulticastJobs() create job of multicast api 1`] = `
Array [
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 0,
        },
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
        Object {
          "id": 4,
        },
      ],
      "to": Array [
        "foo",
        "bar",
        "baz",
      ],
    },
    "entry": "v2/bot/message/multicast",
    "method": "POST",
  },
  Object {
    "body": Object {
      "messages": Array [
        Object {
          "id": 5,
        },
        Object {
          "id": 6,
        },
        Object {
          "id": 7,
        },
      ],
      "to": Array [
        "foo",
        "bar",
        "baz",
      ],
    },
    "entry": "v2/bot/message/multicast",
    "method": "POST",
  },
]
`;
